src/util/file-filter.js

```diff
@@ -0,0 +1,109 @@
+/* @flow */
+import path from 'path';
+
+import minimatch from 'minimatch';
+
+import {createLogger} from './logger';
+
+const log = createLogger(__filename);
+
+// Use this function to mimic path.resolve without resolving to absolute path.
+export const normalizeResolve = (file: string): string => {
+  // normalize
+  file = path.normalize(file);
+
+  // trim trailing slash
+  if (path.parse(file).base && file.endsWith(path.sep)) {
+    return file.slice(0, -1);
+  }
+  return file;
+};
+
+// FileFilter types and implementation.
+
+export type FileFilterOptions = {|
+  filesToIgnore?: Array<string>,
+  ignoreFiles?: Array<string>,
+  sourceDir?: string,
+  artifactsDir?: string,
+|};
+
+/*
+ * Allows or ignores files when creating a ZIP archive.
+ */
+export class FileFilter {
+  filesToIgnore: Array<string>;
+  sourceDir: string;
+
+  constructor({
+    filesToIgnore = [
+      '**/*.xpi',
+      '**/*.zip',
+      '**/.*/', // any hidden file and folder
+      '**/node_modules/',
+    ],
+    ignoreFiles = [],
+    sourceDir = '',
+    artifactsDir,
+  }: FileFilterOptions = {}) {
+
+    this.filesToIgnore = [];
+    this.sourceDir = sourceDir;
+
+    this.addToIgnoreList(filesToIgnore);
+    if (ignoreFiles) {
+      this.addToIgnoreList(ignoreFiles);
+    }
+    if (artifactsDir) {
+      this.addToIgnoreList([artifactsDir], true);
+    }
+  }
+
+  /**
+   *  Resolve relative path to absolute path if sourceDir is setted.
+   */
+  resolve(file: string): string {
+    if (this.sourceDir) {
+      return path.resolve(this.sourceDir, file);
+    }
+    return normalizeResolve(file);
+  }
+
+  /**
+   *  Insert more files into filesToIgnore array.
+   */
+  addToIgnoreList(files: Array<string>, isDir: boolean = false) {
+    for (const file of files) {
+      this.filesToIgnore.push(this.resolve(file));
+      // If file is a directory, ignore its content too.
+      if (/([/\\]\.{0,2})$/.test(file) || isDir) {
+        this.filesToIgnore.push(this.resolve(path.join(file, '**/*')));
+      }
+    }
+  }
+
+  /*
+   * Returns true if the file is wanted for the ZIP archive.
+   *
+   * This is called by zipdir as wantFile(path, stat) for each
+   * file in the folder that is being archived.
+   */
+  wantFile(path: string): boolean {
+    path = this.resolve(path);
```

Comment:
I'm not sure what is the problem described here. `matchBase: true` changes the behavior of match pattern that doesn't contain slashes, make `*.js` behaves like `**/*.js`.

We normalize the path here is to make sure it can match normalized patterns, so `fileFilter.wantFile("path/to/file")` works correctly on Windows. Normally the path is already normalized since `zip-dir`, `watchpack`, and `addons-linter` always pass back full, normalized path, so this change only affect our `test.file-filter.js`.